File: 0001-0500/0036/0036.valid-sudoku.py
#
# @lc app=leetcode id=36 lang=python3
#
# [36] Valid Sudoku
#

# @lc code=start
class Solution:
    def isValidSudoku(self, board: List[List[str]]) -> bool:
        # check row
        for row in board:
            if not self.check(row):
                return False

        # check column
        for i in range(9):
            if not self.check([row[i] for row in board]):
                return False

        # check 3x3
        for i in range(0, 9, 3):
            for j in range(0, 9, 3):
                if not self.check([board[i + x][j + y] for x in range(3) for y in range(3)]):
                    return False

        return True

    def check(self, row):
        row = [x for x in row if x != '.']
        return len(set(row)) == len(row)
# @lc code=end

File: 0001-0500/0036/0036.valid-sudoku.cs
/*
 * @lc app=leetcode id=36 lang=csharp
 *
 * [36] Valid Sudoku
 */

// @lc code=start
public class Solution {
    public bool IsValidSudoku(char[][] board) {
        // Check rows
        for (int i = 0; i < 9; i++) {
            if (!IsValid(board[i])) {
                return false;
            }
        }
        // Check columns
        for (int i = 0; i < 9; i++) {
            char[] column = new char[9];
            for (int j = 0; j < 9; j++) {
                column[j] = board[j][i];
            }
            if (!IsValid(column)) {
                return false;
            }
        }
        // Check 3x3 squares
        for (int i = 0; i < 9; i += 3) {
            for (int j = 0; j < 9; j += 3) {
                char[] square = new char[9];
                int k = 0;
                for (int x = i; x < i + 3; x++) {
                    for (int y = j; y < j + 3; y++) {
                        square[k++] = board[x][y];
                    }
                }
                if (!IsValid(square)) {
                    return false;
                }
            }
        }
        return true;
    }

    private bool IsValid(char[] chars) {
        HashSet<char> set = new HashSet<char>();
        foreach (char c in chars) {
            if (c == '.') {
                continue;
            }
            if (set.Contains(c)) {
                return false;
            }
            set.Add(c);
        }
        return true;
    }
}
// @lc code=end

